crates/core/tedge_api/src/entity_store.rs

```diff
@@ -0,0 +1,539 @@
+    pub fn update(&mut self, message: EntityRegistrationMessage) -> Result<Vec<EntityId>, Error> {
+        if message.r#type == "device" && message.mqtt_id != self.main_device {
+            return Err(Error::MainDeviceAlreadyRegistered(
+                self.main_device.as_str().into(),
+            ));
+        }
+
+        let mut affected_entities = vec![];
+
+        let parent = match message.r#type.as_str() {
+            "child-device" => message.parent.or(Some(self.main_device.clone())),
+            "service" => message.parent.or(Some(self.main_device.clone())),
+            _ => None,
+        };
+
+        // parent device is affected if new device is its child
+        if let Some(parent) = &parent {
+            if !self.entities.contains_key(parent) {
+                return Err(Error::NoParent(parent.clone().into_boxed_str()));
+            }
+
+            affected_entities.push(parent.clone());
+        }
+
+        let external_id = self.external_id(&message.mqtt_id);
+        let entity_metadata = EntityMetadata {
+            r#type: message.r#type,
+            external_id: external_id.clone(),
+            parent,
+            other: message.payload,
+        };
+
+        // device is affected if it was previously registered and was updated
+        let previous = self
+            .entities
+            .insert(message.mqtt_id.clone(), entity_metadata);
+
+        if previous.is_some() {
+            affected_entities.push(message.mqtt_id);
+        } else {
+            self.external_id_index.insert(external_id, message.mqtt_id);
+        }
+
+        Ok(affected_entities)
+    }
+
+    /// An iterator over all registered entities.
+    pub fn iter(&self) -> impl Iterator<Item = (&EntityId, &EntityMetadata)> {
+        self.entities.iter()
+    }
+
+    /// Generate child device external ID.
+    ///
+    /// The external id is generated by prefixing the id with main device name
+    /// (device_common_name) and then appending the MQTT entity topic with `/`
+    ///  characters replaced by `:`.
+    ///
+    /// # Examples
+    /// - `device/main//` => `DEVICE_COMMON_NAME`
+    /// - `device/child001//` => `DEVICE_COMMON_NAME:device:child001`
+    /// - `device/child001/service/service001` => `DEVICE_COMMON_NAME:device:child001:service:service001`
+    /// - `factory01/hallA/packaging/belt001` => `DEVICE_COMMON_NAME:factory01:hallA:packaging:belt001`
+    fn external_id(&self, entity_topic: EntityIdRef) -> String {
+        let main_device_external_id = &self.get(&self.main_device).unwrap().external_id;
+        let external_id_suffix = entity_topic.replace('/', ":");
+        let external_id_suffix = external_id_suffix.trim_matches(':');
+        format!("{main_device_external_id}:{external_id_suffix}")
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct EntityMetadata {
+    parent: Option<EntityId>,
+    r#type: String,
+    external_id: String,
+    other: serde_json::Value,
+}
```

Comment:
I would consider to use an enum here, ruling out the cases that makes no sense.

```suggestion
#[derive(Debug, Clone)]
pub struct EntityMetadata {
    r#type: EntityType,
    other: serde_json::Value,
}

pub enum EntityType {
    MainDevice{ device_id: String },
    ChildDevice( device_id: String, parent: EntityId },
    Service( service_id: String, device: EntityId },
}
```

Reply:
Because `device_id` and `service_id` both refer to the external id, and in the case of child devices and services `parent` and `device` mean basically the same thing, I've opted to make the entity type just a plain enum.